From 6a300a4ed1c1864a20683bab345bf5d992280240 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Fri, 2 Dec 2011 13:15:58 -0700
Subject: [PATCH] Implement service to assign changes to users.

This change merely implements the service to assign
changes to users, it does not introduce any users of
the service.

Change-Id: Ieeaf497bcc9bf92f92731d0308786c6487a04db9
---
 .../gerrit/common/data/ChangeManageService.java    |    5 +
 .../httpd/rpc/changedetail/AssignChange.java       |   91 ++++++++++++++++++++
 .../rpc/changedetail/ChangeManageServiceImpl.java  |   13 +++-
 .../httpd/rpc/changedetail/ChangeModule.java       |    1 +
 .../java/com/google/gerrit/server/ChangeUtil.java  |    5 +-
 5 files changed, 111 insertions(+), 4 deletions(-)
 create mode 100644 gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/AssignChange.java

diff --git a/gerrit-common/src/main/java/com/google/gerrit/common/data/ChangeManageService.java b/gerrit-common/src/main/java/com/google/gerrit/common/data/ChangeManageService.java
index b667877..54031ea 100644
--- a/gerrit-common/src/main/java/com/google/gerrit/common/data/ChangeManageService.java
+++ b/gerrit-common/src/main/java/com/google/gerrit/common/data/ChangeManageService.java
@@ -15,6 +15,7 @@
 package com.google.gerrit.common.data;
 
 import com.google.gerrit.common.auth.SignInRequired;
+import com.google.gerrit.reviewdb.Account;
 import com.google.gerrit.reviewdb.PatchSet;
 import com.google.gwt.user.client.rpc.AsyncCallback;
 import com.google.gwtjsonrpc.client.RemoteJsonService;
@@ -44,4 +45,8 @@ public interface ChangeManageService extends RemoteJsonService {
 
   @SignInRequired
   void deleteDraftChange(PatchSet.Id patchSetId, AsyncCallback<VoidResult> callback);
+
+  @SignInRequired
+  void assignChange(PatchSet.Id patchSetId, final Account.Id toId,
+      String message, AsyncCallback<ChangeDetail> callback);
 }
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/AssignChange.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/AssignChange.java
new file mode 100644
index 0000000..682f93f
--- /dev/null
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/AssignChange.java
@@ -0,0 +1,91 @@
+// Copyright (C) 2012 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.gerrit.httpd.rpc.changedetail;
+
+import com.google.gerrit.common.ChangeHookRunner;
+import com.google.gerrit.common.data.ChangeDetail;
+import com.google.gerrit.common.errors.NoSuchEntityException;
+import com.google.gerrit.httpd.rpc.Handler;
+import com.google.gerrit.reviewdb.Account;
+import com.google.gerrit.reviewdb.Change;
+import com.google.gerrit.reviewdb.PatchSet;
+import com.google.gerrit.reviewdb.ReviewDb;
+import com.google.gerrit.server.ChangeUtil;
+import com.google.gerrit.server.IdentifiedUser;
+import com.google.gerrit.server.mail.EmailException;
+import com.google.gerrit.server.mail.RestoredSender;
+import com.google.gerrit.server.patch.PatchSetInfoNotAvailableException;
+import com.google.gerrit.server.project.ChangeControl;
+import com.google.gerrit.server.project.InvalidChangeOperationException;
+import com.google.gerrit.server.project.NoSuchChangeException;
+import com.google.gwtorm.client.OrmException;
+import com.google.inject.Inject;
+import com.google.inject.assistedinject.Assisted;
+
+import javax.annotation.Nullable;
+
+class AssignChange extends Handler<ChangeDetail> {
+  interface Factory {
+    AssignChange create(PatchSet.Id patchSetId, final Account.Id toId, String message);
+  }
+
+  private final ChangeControl.Factory changeControlFactory;
+  private final ReviewDb db;
+  private final IdentifiedUser currentUser;
+  private final IdentifiedUser.GenericFactory identifiedUserFactory;
+  private final ChangeDetailFactory.Factory changeDetailFactory;
+
+  private final PatchSet.Id patchSetId;
+  private final Account.Id toId;
+  @Nullable
+  private final String message;
+
+  @Inject
+  AssignChange(final ChangeControl.Factory changeControlFactory,
+      final ReviewDb db, final IdentifiedUser currentUser,
+      final IdentifiedUser.GenericFactory identifiedUserFactory,
+      final ChangeDetailFactory.Factory changeDetailFactory,
+      @Assisted final PatchSet.Id patchSetId,
+      @Assisted @Nullable final Account.Id toId,
+      @Assisted @Nullable final String message) {
+    this.changeControlFactory = changeControlFactory;
+    this.db = db;
+    this.currentUser = currentUser;
+    this.identifiedUserFactory = identifiedUserFactory;
+    this.changeDetailFactory = changeDetailFactory;
+
+    this.patchSetId = patchSetId;
+    this.toId = toId;
+    this.message = message;
+  }
+
+  @Override
+  public ChangeDetail call() throws NoSuchChangeException, OrmException,
+      EmailException, NoSuchEntityException, InvalidChangeOperationException,
+      PatchSetInfoNotAvailableException {
+    final Change.Id changeId = patchSetId.getParentKey();
+    final ChangeControl control = changeControlFactory.validateFor(changeId);
+
+    final IdentifiedUser toUser = identifiedUserFactory.create(toId);
+    if (control.forUser(toUser).isRefVisible()) {
+      throw new InvalidChangeOperationException("Cannot assign change since it"
+          + " is not visible to " + toUser.getName());
+    }
+
+    ChangeUtil.assign(patchSetId, currentUser, toId, message, db);
+
+    return changeDetailFactory.create(changeId).call();
+  }
+}
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeManageServiceImpl.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeManageServiceImpl.java
index 11434a7..9318365 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeManageServiceImpl.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeManageServiceImpl.java
@@ -16,6 +16,7 @@ package com.google.gerrit.httpd.rpc.changedetail;
 
 import com.google.gerrit.common.data.ChangeDetail;
 import com.google.gerrit.common.data.ChangeManageService;
+import com.google.gerrit.reviewdb.Account;
 import com.google.gerrit.reviewdb.PatchSet;
 import com.google.gwt.user.client.rpc.AsyncCallback;
 import com.google.gwtjsonrpc.client.VoidResult;
@@ -28,6 +29,7 @@ class ChangeManageServiceImpl implements ChangeManageService {
   private final RevertChange.Factory revertChangeFactory;
   private final PublishAction.Factory publishAction;
   private final DeleteDraftChange.Factory deleteDraftChangeFactory;
+  private final AssignChange.Factory assignChangeFactory;
 
   @Inject
   ChangeManageServiceImpl(final SubmitAction.Factory patchSetAction,
@@ -35,13 +37,15 @@ class ChangeManageServiceImpl implements ChangeManageService {
       final RestoreChange.Factory restoreChangeFactory,
       final RevertChange.Factory revertChangeFactory,
       final PublishAction.Factory publishAction,
-      final DeleteDraftChange.Factory deleteDraftChangeFactory) {
+      final DeleteDraftChange.Factory deleteDraftChangeFactory,
+      final AssignChange.Factory assignChangeFactory) {
     this.submitAction = patchSetAction;
     this.abandonChangeFactory = abandonChangeFactory;
     this.restoreChangeFactory = restoreChangeFactory;
     this.revertChangeFactory = revertChangeFactory;
     this.publishAction = publishAction;
     this.deleteDraftChangeFactory = deleteDraftChangeFactory;
+    this.assignChangeFactory = assignChangeFactory;
   }
 
   public void submit(final PatchSet.Id patchSetId,
@@ -73,4 +77,9 @@ class ChangeManageServiceImpl implements ChangeManageService {
       final AsyncCallback<VoidResult> callback) {
     deleteDraftChangeFactory.create(patchSetId).to(callback);
   }
-}
\ No newline at end of file
+
+  public void assignChange(final PatchSet.Id patchSetId, final Account.Id toId,
+      final String message, final AsyncCallback<ChangeDetail> callback) {
+    assignChangeFactory.create(patchSetId, toId, message).to(callback);
+  }
+}
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeModule.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeModule.java
index 34685ec..0d5ff9b 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeModule.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeModule.java
@@ -28,6 +28,7 @@ public class ChangeModule extends RpcServletModule {
     install(new FactoryModule() {
       @Override
       protected void configure() {
+        factory(AssignChange.Factory.class);
         factory(AbandonChange.Factory.class);
         factory(RestoreChange.Factory.class);
         factory(RevertChange.Factory.class);
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/ChangeUtil.java b/gerrit-server/src/main/java/com/google/gerrit/server/ChangeUtil.java
index 7ad22a3..25c0008 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/ChangeUtil.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/ChangeUtil.java
@@ -532,13 +532,14 @@ public class ChangeUtil {
     db.patchSets().delete(Collections.singleton(patch));
   }
 
-  public static void assign(final Change.Id changeId, final IdentifiedUser user,
+  public static void assign(final PatchSet.Id patchSetId, final IdentifiedUser user,
       final Account.Id toId, final String message, final ReviewDb db)
       throws NoSuchChangeException, InvalidChangeOperationException,
       OrmException {
+    final Change.Id changeId = patchSetId.getParentKey();
     final ChangeMessage cmsg =
         new ChangeMessage(new ChangeMessage.Key(changeId, ChangeUtil
-            .messageUUID(db)), user.getAccountId());
+            .messageUUID(db)), user.getAccountId(), patchSetId);
 
     final StringBuilder msgBuf = new StringBuilder();
     if (toId != null) {
-- 
1.7.8

